server.go

```diff
@@ -862,6 +863,120 @@ func (sp *serverPeer) OnMiningState(p *peer.Peer, msg *wire.MsgMiningState) {
 	}
 }
 
+// OnGetInitState is invoked when a peer receives a getinitstate wire message.
+// It sends the available requested info the the remote peer.
+func (sp *serverPeer) OnGetInitState(p *peer.Peer, msg *wire.MsgGetInitState) {
+	if sp.initStateSent {
+		peerLog.Tracef("Ignoring getinitstate from %v - already sent", sp.Peer)
+		return
+	}
+	sp.initStateSent = true
+
+	// Access the block manager and get the list of best blocks to mine on.
+	bm := sp.server.blockManager
+	mp := sp.server.txMemPool
+	best := sp.server.chain.BestSnapshot()
+
+	// Send out blank mining states if it's early in the blockchain.
+	if best.Height < sp.server.chainParams.StakeValidationHeight-1 {
+		sp.QueueMessage(wire.NewMsgInitState(), nil)
+		return
+	}
+
+	// Response data.
+	var blockHashes, voteHashes, tspendHashes []chainhash.Hash
+
+	// Map from the types slice into a map for easier checking.
+	types := make(map[string]struct{}, len(msg.Types))
+	for _, typ := range msg.Types {
+		types[typ] = struct{}{}
+	}
+	_, wantBlocks := types[wire.InitStateHeadBlocks]
+	_, wantVotes := types[wire.InitStateHeadBlockVotes]
+	_, wantTSpends := types[wire.InitStateTSpends]
+
+	// Fetch head block hashes if we need to send either them or their
+	// votes.
+	if wantBlocks || wantVotes {
+		// Obtain the entire generation of blocks stemming from the
+		// parent of the current tip.
+		children, err := bm.TipGeneration()
+		if err != nil {
+			peerLog.Warnf("Failed to access block manager to get the generation "+
+				"for a init state request (block: %v): %v", best.Hash, err)
+			return
+		}
+
+		// Get the list of blocks that are eligible to build on and
+		// limit the list to the maximum number of allowed eligible
+		// block hashes per mining state message.  There is nothing to
+		// send when there are no eligible blocks.
+		blockHashes = mining.SortParentsByVotes(mp, best.Hash, children,
+			bm.cfg.ChainParams)
+		if len(blockHashes) > wire.MaxISBlocksAtHeadPerMsg {
+			blockHashes = blockHashes[:wire.MaxISBlocksAtHeadPerMsg]
+		}
+	}
+
+	// Construct the set of votes to send.
+	if wantVotes {
+		for i := range blockHashes {
+			// Fetch the vote hashes themselves and append them.
+			bh := &blockHashes[i]
+			vhsForBlock := mp.VoteHashesForBlock(bh)
+			if len(vhsForBlock) == 0 {
+				peerLog.Warnf("Unexpected error while fetching vote hashes "+
+					"for block %v for an init state request: no vote "+
+					"metadata for block", bh)
+				return
```

Comment:
Also, for the same reason, I think this should be a continue, so it can deliver the votes it does know about.  This would cause no response to be sent despite there might having been tspends requested and available (or indeed other block hashes and vote hashes).